File: Lista/Go/Q11.go
package main

import "fmt"

type Funcionario interface {
    CalcularSalario() float64
}

type FuncionarioHorista struct {
    Nome       string
    HorasTrabalhadas float64
    ValorHora  float64
}

func (fh FuncionarioHorista) CalcularSalario() float64 {
    return fh.HorasTrabalhadas * fh.ValorHora
}

type FuncionarioAssalariado struct {
    Nome         string
    SalarioMensal float64
}

func (fa FuncionarioAssalariado) CalcularSalario() float64 {
    return fa.SalarioMensal
}

func ExibirSalario(f Funcionario) {
    fmt.Printf("Salário: R$%.2f\n", f.CalcularSalario())
}

func main() {
    funcionarioHorista := FuncionarioHorista{
        Nome:          "Alice",
        HorasTrabalhadas: 160,
        ValorHora:     50.0,
    }

    funcionarioAssalariado := FuncionarioAssalariado{
        Nome:         "Bob",
        SalarioMensal: 5000.0,
    }

    fmt.Println("Funcionário Horista:")
    ExibirSalario(funcionarioHorista)

    fmt.Println("Funcionário Assalariado:")
    ExibirSalario(funcionarioAssalariado)
}

File: Lista/Java/Q11.java
abstract class Funcionario {
    protected String nome;

    public Funcionario(String nome) {
        this.nome = nome;
    }

    public abstract double calcularSalario();
}

class FuncionarioHorista extends Funcionario {
    private double valorHora;
    private int horasTrabalhadas;

    public FuncionarioHorista(String nome, double valorHora, int horasTrabalhadas) {
        super(nome);
        this.valorHora = valorHora;
        this.horasTrabalhadas = horasTrabalhadas;
    }

    @Override
    public double calcularSalario() {
        return valorHora * horasTrabalhadas;
    }
}

class FuncionarioAssalariado extends Funcionario {
    private double salarioMensal;

    public FuncionarioAssalariado(String nome, double salarioMensal) {
        super(nome);
        this.salarioMensal = salarioMensal;
    }

    @Override
    public double calcularSalario() {
        return salarioMensal;
    }
}

public class Main {
    public static void main(String[] args) {
        Funcionario horista = new FuncionarioHorista("João", 20.0, 40);
        System.out.println("Funcionário Horista: " + horista.nome + ", Salário: R$ " + horista.calcularSalario());

        Funcionario assalariado = new FuncionarioAssalariado("Maria", 3000.0);
        System.out.println("Funcionário Assalariado: " + assalariado.nome + ", Salário: R$ " + assalariado.calcularSalario());
    }
}
